### ISSUE TITLE
Cache utils produces a graph compilation for each tensor in the model

### ISSUE THREAD

**Issue by aws-rhsoln** [huggingface/optimum-neuron#301]
There is an issue with the NeuronCache class. It tries to compute the hash of the model by copying each tensor one by one to CPU. This would cause a graph compile for each tensor: https://github.com/huggingface/optimum-neuron/blob/main/optimum/neuron/utils/cache_utils.py#L632 .

**Comment by michaelbenayoun** [comment]
It has been like that since the beginning. Is that a bad thing?
Would you have any suggestion on another way to get the content of each tensor to compute a hash?

**Comment by aws-rhsoln** [comment]
Yeah, its not going to be great since it would cause a compile for each tensor. So one recommendation would be to have a mark_step to materialize all the tensors and then you can copy the tensors one by one. This would not incur a compile. Btw, is this just a one time thing? Since copying tensors to CPU is expensive.

**Comment by michaelbenayoun** [comment]
Interesting. I will add the `xm.mark_step()` before doing that.
It's roughly a one-time thing. Basically it should ahppen again if we encounter a different model.
